### ISSUE TITLE
foward-to-forward network connections cause a huge memory leak on the server side

### ISSUE THREAD

**Issue by cazorla19** [fluent/fluent-bit#9093]
## Bug Report

**Describe the bug**

We're running a log pipeline system based on fluent-bit where 2 log aggregators on datacenter-1 send logs to 2 log aggregators on datacenter-2 via `forward` plugin using the `upstream` feature and TLS verification. For some reason, the way connections are being handled causes a huge memory leak on the server side, which, in this context, are datacenter-2 servers. The RAM usage gets constantly increased by ca. 350 MB per hour. We've had to set up the memory limits in systemd, otherwise the 8GB node would go down after a day of running fluent-bit.

This issue doesn't occur when there is no incoming traffic on the server-side.

We tried setting up `net.keepalive` on the client side with `net.keepalive_idle_timeout = 10`, and it kind of helped reducing a memory growth on the server side, but after the client restart the of memory consumption got back to its scale as nothing has been changed. Which is even more misterious, because why would this change fix the problem, and then the problem would come back again as if we've changed nothing.

Below I've attached the config samples together with Valgrind output we've used to identify issues running the same fluent-bit version compiled with ` CMAKE_BUILD_TYPE=debug`.

**To Reproduce**

Run 2 fluent-bit clients (on Windows) and 2 fluent-bit servers (on Linux) connected via `forward` plugin using the `upstream` feature and TLS verification.

- Example log message if applicable:

Logs don't tell us anything useful even in debug mode, however we've run fluent-bit binary compiled with ` CMAKE_BUILD_TYPE=debug` via Valgrind, and here is what we've got after running it for 40 minutes:

```
==8530== LEAK SUMMARY:
==8530==    definitely lost: 111,046,732 bytes in 1,524 blocks
==8530==    indirectly lost: 260,390,021 bytes in 5,371,533 blocks
==8530==      possibly lost: 23,451,836 bytes in 530,281 blocks
==8530==    still reachable: 446 bytes in 8 blocks
==8530==         suppressed: 0 bytes in 0 blocks
==8530==
==8530== For lists of detected and suppressed errors, rerun with: -s
==8530== ERROR SUMMARY: 25 errors from 25 contexts (suppressed: 0 from 0)
```

Here is the complete report.

[valgrind.txt](https://github.com/user-attachments/files/16244886/valgrind.txt)

**Expected behavior**

foward-to-forward connection works without a huge memory leak on the server side, and the memory usage gets increased only when the log pipeline throughput gets increased.

**Screenshots**

![image](https://github.com/user-attachments/assets/38e707fe-545d-493a-aa9e-18a1b73f60d3)

**Your Environment**

* Version used: 3.0.7-windows (forward-clients) & 3.0.7-linux (forward-servers)
* Configuration:

Clients:

```
# fluent-bit.conf

﻿[SERVICE] # windows
  flush 1
  daemon false
  log_level info
  parsers_file parsers.conf
  plugins_file plugins.conf
  http_server On
  http_listen 0.0.0.0
  http_port 2020
  storage.max_chunks_up 10
  storage.delete_irrecoverable_chunks true
  storage.metrics true
  log_file C:\Program Files\fluent-bit\fluent-bit-log.txt
  storage.path E:\temp\fluent-bit

[INPUT]
  name winlog
  Channels System, Security
  Interval Sec 1
  DB winlog.sqlite

[OUTPUT]
  name forward 
  Match *
  Upstream upstream.conf
  Retry_Limit no_limits

# upstream.conf

[UPSTREAM]
  name forward-balancing

[NODE]
  name node1
  Host node1.example.com
  Port 24224
  tls on
  tls.verify on
  net.keepalive on
  net.keepalive_idle_timeout 10

[NODE]
  name node2
  Host node2.example.com
  Port 24224
  tls on
  tls.verify on
  net.keepalive on
  net.keepalive_idle_timeout 10
```

Server:

```
# fluent-bit.conf

[SERVICE]
# linux
    flush  1
    daemon  false
    log_level  info
    http_server  On
    http_listen  0.0.0.0
    http_port  2020
    storage.max_chunks_up  10
    storage.delete_irrecoverable_chunks  true
    storage.metrics  true
    storage.path  /var/cache/fluent-bit

[INPUT]
    name forward
    Mem_Buf_Limit  10MB
    storage.type  filesystem
    Listen  0.0.0.0
    Port  24224
    tls  true
    tls.verify  true
    tls.crt_file  /etc/certs/fullchain.pem
    tls.key_file  /etc/certs/privkey.pem

# the output plugins could somehow cause this issue, but I'm not sure of it
[OUTPUT]
    name loki
    host  loki.example.com
    port  443
    tls  On
    line_format  key_value
    match  *

[OUTPUT]
    name s3
    bucket  example
    match  *
    compression  gzip
    upload_timeout  1h
    use_put_object  true
    total_file_size  500M
```

* Operating System and version:

- clients are running on Windows Server 2019 VMs as windows services
- servers are running on Oracle Linux 8.10 VMs as systemd services

* Filters and plugins:

- `forward`
- `s3` and `loki` as the final endpoints

**Additional context**

This issue in our context is critical, because with that huge memory leak the system gets completely unmaitainable. We'd also try to use `http` plugin instead of the `forward` one, but we're not exactly sure why `forward` works that way.

**Comment by cosmo0920** [comment]
I diagnosed this issue and I found that most of this issue is originated from out_s3 plugin. This is because:

```
==8530== 110,926,948 bytes in 21 blocks are definitely lost in loss record 46 of 47
==8530==    at 0x4C39185: malloc (vg_replace_malloc.c:442)
==8530==    by 0xF0898D: cio_file_content_copy (in /opt/fluent-bit/code/fluent-bit/bin/fluent-bit)
==8530==    by 0xF05568: cio_chunk_get_content_copy (in /opt/fluent-bit/code/fluent-bit/bin/fluent-bit)
==8530==    by 0xB82CE3: flb_fstore_file_content_copy (in /opt/fluent-bit/code/fluent-bit/bin/fluent-bit)
==8530==    by 0xA7F976: s3_store_file_read (in /opt/fluent-bit/code/fluent-bit/bin/fluent-bit)
==8530==    by 0xA76D08: construct_request_buffer (in /opt/fluent-bit/code/fluent-bit/bin/fluent-bit)
==8530==    by 0xA7688B: put_all_chunks (in /opt/fluent-bit/code/fluent-bit/bin/fluent-bit)
==8530==    by 0xA7AB85: cb_s3_exit (in /opt/fluent-bit/code/fluent-bit/bin/fluent-bit)
==8530==    by 0x54F705: flb_output_exit (in /opt/fluent-bit/code/fluent-bit/bin/fluent-bit)
==8530==    by 0x577545: flb_engine_shutdown (in /opt/fluent-bit/code/fluent-bit/bin/fluent-bit)
==8530==    by 0x57717B: flb_engine_start (in /opt/fluent-bit/code/fluent-bit/bin/fluent-bit)
==8530==    by 0x518F12: flb_lib_worker (in /opt/fluent-bit/code/fluent-bit/bin/fluent-bit)
```

```
irb> ((111_046_732  - 110_926_948 ) / 111_046_732.0 ) * 100
=> 0.1078681000715987
```

Around 99.9 % leaks are originated from out_s3.

**Comment by edsiper** [comment]
Fixed by #9143
